File: sample-app-java/app/src/main/java/com/bluejeans/android/sdksample/isc/IscGalleryViewModel.java
package com.bluejeans.android.sdksample.isc;

import android.util.Log;
import android.view.TextureView;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.bjnclientcore.media.individualstream.VideoStreamConfigurationResult;
import com.bluejeans.android.sdksample.SampleApplication;
import com.bluejeans.bluejeanssdk.meeting.ParticipantsService;
import com.bluejeans.bluejeanssdk.meeting.individualstream.VideoStreamService;
import com.bluejeans.rxextensions.ObservableComputed;
import com.bluejeans.rxextensions.ObservableValue;
import com.bluejeans.rxextensions.ObservableVariable;
import org.jetbrains.annotations.NotNull;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;
import java.util.stream.Collectors;

import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.schedulers.Schedulers;
import kotlin.collections.MapsKt;
import kotlin.jvm.internal.Intrinsics;
import timber.log.Timber;

public class IscGalleryViewModel extends ViewModel {
    private final String TAG = "IscGalleryViewModel";
    private final VideoStreamService videoStreamService;
    private final ParticipantsService participantService;
    private final CompositeDisposable compositeDisposable;
    private final MutableLiveData<List<ParticipantUI>> _participantUIObservable;
    private final MutableLiveData _rosterParticipantsObservable;
    private final ObservableVariable<Map> requestedQuality;
    private final ObservableVariable<Map<String, Boolean>> pinnedParticipants;
    private final ObservableVariable<Map<String, ParticipantsService.Participant>> bjnParticipantMap;
    private final LiveData rosterParticipantsObservable;
    private final LiveData<List<ParticipantUI>> participantUIObservable;
    private final ExecutorService schedulers = Executors.newSingleThreadExecutor();

    private ObservableValue<List<ParticipantUI>> individualStreamsObserver;

    public IscGalleryViewModel() {
        this.videoStreamService = SampleApplication.getBlueJeansSDK().getMeetingService().getVideoStreamService();
        this.participantService = SampleApplication.getBlueJeansSDK().getMeetingService().getParticipantsService();
        compositeDisposable = new CompositeDisposable();
        _participantUIObservable = new MutableLiveData<List<ParticipantUI>>();
        _rosterParticipantsObservable = new MutableLiveData();
        this.requestedQuality = new ObservableVariable<Map>((new LinkedHashMap()), false);
        this.pinnedParticipants = new ObservableVariable<Map<String, Boolean>>((new LinkedHashMap<String, Boolean>()), false);
        this.bjnParticipantMap = new ObservableVariable<>(MapsKt.emptyMap(), false);
        this.rosterParticipantsObservable = this._rosterParticipantsObservable;
        this.participantUIObservable = this._participantUIObservable;
        subscribeToIndividualStreams();
    }

    @NotNull
    public final LiveData<List<ParticipantUI>> getParticipantUIObservable() {
        return this.participantUIObservable;
    }

    @NotNull
    public final VideoStreamConfigurationResult setVideoConfiguration(List<com.bjnclientcore.media.individualstream.VideoStreamConfiguration> list) {
        Intrinsics.checkNotNullParameter(list, "list");
        return this.videoStreamService.setVideoStreamConfiguration(list);
    }

    public final void attachParticipantToView(@NotNull String participantGuid, @NotNull TextureView textureView) {
        Intrinsics.checkNotNullParameter(participantGuid, "participantGuid");
        Intrinsics.checkNotNullParameter(textureView, "textureView");
        this.videoStreamService.attachParticipantToView(participantGuid, textureView);
    }

    public final void detachParticipantFromView(@NotNull String participantGuid) {
        Intrinsics.checkNotNullParameter(participantGuid, "participantGuid");
        this.videoStreamService.detachParticipantFromView(participantGuid);
    }

    public final void pinParticipant(@NotNull String id) {
        Intrinsics.checkNotNullParameter(id, "id");
        Map<String, Boolean> currentParticipants = this.pinnedParticipants.getValue();
        currentParticipants.put(id, true);
        this.pinnedParticipants.setValue(currentParticipants);
    }

    public final void unpinParticipant(@NotNull String id) {
        Intrinsics.checkNotNullParameter(id, "id");
        Map<String, Boolean> currentParticipants = this.pinnedParticipants.getValue();
        currentParticipants.remove(id);
        this.pinnedParticipants.setValue(currentParticipants);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        compositeDisposable.clear();
    }

    private void subscribeToIndividualStreams() {
        compositeDisposable.add(
                participantService.getParticipants().getRxObservable()
                        .filter(f -> f.getValue() != null)
                        .subscribeOn(Schedulers.from(schedulers))
                        .subscribe(participants -> {
                            Map<String, ParticipantsService.Participant> participantMap = participants.getValue().stream().filter(p -> !p.getId().equals(participantService.getSelfParticipant().getValue().getId()))
                                    .collect(Collectors.toMap(ParticipantsService.Participant::getId, Function.identity()));
                            bjnParticipantMap.setValue(
                                    participantMap
                            );

                        }, err -> {
                            Log.e(TAG, "Error in getting roster updates: " + err.getMessage());
                        })
        );

        individualStreamsObserver = ObservableComputed.Companion.create(bjnParticipantMap.readonly(), videoStreamService.getVideoStreams(), false, (rosterParticipantMap, videoStream) -> {
            List<ParticipantUI> videoParticipants = new ArrayList<>();
            List<ParticipantUI> nonVideoParticipants = new ArrayList<>();

            rosterParticipantMap.forEach((key, value) -> {
                if (videoStream.stream().anyMatch(s -> s.getParticipantGuid().equals(key))) {
                    videoStream.forEach(stream -> {
                        if (stream.getParticipantGuid().equals(key)) {
                            Log.i(TAG, "Adding stream for " + value.getId() + " and name " + value.getName());
                            int width = 0, height = 0;
                            if (stream.getVideoResolution().getValue() != null) {
                                width = stream.getVideoResolution().getValue().getWidth();
                                height = stream.getVideoResolution().getValue().getHeight();
                            }
                            videoParticipants.add(new ParticipantUI(
                                    value.getId(), value.getName(),
                                    width,
                                    height,
                                    stream.getVideoResolution(),
                                    true,
                                    false
                            ));
                        }
                    });

                } else {
                    Log.i(TAG, "No stream found for " + value.getId() + " and name " + value.getName());
                    nonVideoParticipants.add(new ParticipantUI(
                            value.getId(), value.getName(),
                            0,
                            0,
                            null,
                            false,
                            false
                    ));
                }
            });

            ArrayList<ParticipantUI> allParticipants = new ArrayList<>();
            videoParticipants.forEach(p -> allParticipants.add(p));
            nonVideoParticipants.forEach(p -> allParticipants.add(p));

            return allParticipants;
        });

        compositeDisposable.add(individualStreamsObserver.getRxObservable()
                .debounce(200, TimeUnit.MILLISECONDS, Schedulers.from(schedulers))
                .subscribe(it -> {
                    Timber.tag(TAG).i("Roster participants fetched size " + it.size());
                    int finalIndex = Math.min(it.size(), 25);
                    _participantUIObservable.postValue(it.subList(0, finalIndex));
                }, err -> {
                    Log.e(TAG, "Error in getting streams: " + err.getMessage());
                }));
    }
}

File: sample-app-kotlin/app/src/main/java/com/bluejeans/android/sdksample/isc/IscGalleryViewModel.kt
package com.bluejeans.android.sdksample.isc

import android.util.Size
import android.view.TextureView
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.bjnclientcore.media.individualstream.DetachParticipantStreamResult
import com.bjnclientcore.media.individualstream.StreamPriority
import com.bjnclientcore.media.individualstream.StreamQuality
import com.bjnclientcore.media.individualstream.VideoStreamConfiguration
import com.bjnclientcore.media.individualstream.VideoStreamConfigurationResult
import com.bluejeans.android.sdksample.SampleApplication
import com.bluejeans.bluejeanssdk.meeting.ParticipantsService
import com.bluejeans.rxextensions.ObservableComputed
import com.bluejeans.rxextensions.ObservableValueWithOptional
import com.bluejeans.rxextensions.ObservableVariable
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.kotlin.addTo
import io.reactivex.rxjava3.schedulers.Schedulers
import timber.log.Timber
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

data class ParticipantUI(
    val seamGuid: String, var name: String,
    val videoWidth: Int? = 0,
    val videoHeight: Int? = 0,
    val videoResolution: ObservableValueWithOptional<Size>? = null,
    val streamPriority: StreamPriority? = null,
    val streamQuality: StreamQuality? = null,
    val pinned: Boolean = false,
    val isVideo: Boolean
)

class IscGalleryViewModel : ViewModel() {
    private val TAG = "IscGalleryViewModel"

    private val videoStreamService =
        SampleApplication.blueJeansSDK.meetingService.videoStreamService
    private val participantService =
        SampleApplication.blueJeansSDK.meetingService.participantsService

    private val compositeDisposable = CompositeDisposable()

    private val _participantUIObservable = MutableLiveData<List<ParticipantUI>>()

    private val _rosterParticipantsObservable = MutableLiveData<List<ParticipantUI>>()

    private val requestedQuality = ObservableVariable<MutableMap<String, StreamQuality>>(
        mutableMapOf()
    )
    private val pinnedParticipants = ObservableVariable<MutableMap<String, Boolean>>(
        mutableMapOf()
    )

    private val bjnParticipantMap: ObservableVariable<Map<String, ParticipantsService.Participant>> =
        ObservableVariable(
            emptyMap(), false
        )

    private val rosterParticipantsObservable: LiveData<List<ParticipantUI>> =
        _rosterParticipantsObservable

    val participantUIObservable: LiveData<List<ParticipantUI>> = _participantUIObservable

    val schedulers = Executors.newSingleThreadExecutor()

    init {
        participantService.participants
            .rxObservable
            .filter {
                it.value != null
            }
            .subscribeOn(Schedulers.from(schedulers))
            .subscribe({
                val map =
                    it.value!!.filter { it1 -> it1.id != participantService.selfParticipant.value?.id }
                        .associateBy({ it1 -> it1.id }, { it2 -> it2 })
                bjnParticipantMap.value = map
            }, {
                Timber.tag(TAG).e("Error observing ${it.stackTraceToString()}")
            }).addTo(compositeDisposable)

        ObservableComputed.create(
            bjnParticipantMap,
            videoStreamService.videoStreams,
            compositeDisposable,
            distinctUntilChanged = false
        ) { rosterParticipants, videoStream ->
            val videoParticipants = mutableListOf<ParticipantUI>()
            val nonVideoParticipants = mutableListOf<ParticipantUI>()
            Timber.tag(TAG).i("Video Streams size ${videoStream.size} ")
            rosterParticipants.filter { it.key != participantService.selfParticipant.value?.id }
                .forEach { (key, value) ->
                    if (videoStream.find { it.participantGuid == key } != null) {
                        videoStream.distinct().forEach { stream ->
                            if (stream.participantGuid == key) {
                                videoParticipants.add(
                                    ParticipantUI(
                                        value.id,
                                        value.name,
                                        stream.videoResolution.value?.width,
                                        stream.videoResolution.value?.height,
                                        stream.videoResolution,
                                        isVideo = true,
                                        pinned = true
                                    )
                                )
                            }
                        }
                    } else {
                        nonVideoParticipants.add(
                            ParticipantUI(
                                value.id,
                                value.name,
                                pinned = false,
                                isVideo = false
                            )
                        )
                    }
                }
            videoParticipants + nonVideoParticipants
        }
            .rxObservable
            .debounce(200, TimeUnit.MILLISECONDS, Schedulers.from(schedulers))
            .subscribe(
                {
                    Timber.tag(TAG)
                        .i("Roster participants fetched size ${it.size}")
                    _participantUIObservable.postValue(it.take(25))
                },
                {
                    Timber.tag(TAG).e("Error observing ${it.stackTraceToString()}")
                }).addTo(compositeDisposable)
    }


    fun setVideoConfiguration(list: List<VideoStreamConfiguration>): VideoStreamConfigurationResult {
        return videoStreamService.setVideoStreamConfiguration(list)
    }

    fun attachParticipantToView(participantGuid: String, textureView: TextureView) {
        videoStreamService.attachParticipantToView(participantGuid, textureView)
    }

    fun detachParticipantFromView(participantGuid: String): DetachParticipantStreamResult {
        return videoStreamService.detachParticipantFromView(participantGuid)
    }

    fun pinParticipant(id: String) {
        val currentParticipants = pinnedParticipants.value
        currentParticipants[id] = true
        pinnedParticipants.value = currentParticipants
    }

    fun unpinParticipant(id: String) {
        val currentParticipants = pinnedParticipants.value
        currentParticipants.remove(id)
        pinnedParticipants.value = currentParticipants
    }


    override fun onCleared() {
        super.onCleared()
        compositeDisposable.clear()
    }
}